File: swift_analyzer/swift799_analyzer.py
import re
from typing import Optional
from .ml.deberta_model import DeBertaModel
from .models import Swift799Message

class Swift799Analyzer:
    def __init__(self, model_name: str = "MoritzLaurer/mDeBERTa-v3-base-mnli-xnli", extract_narrative_only: bool = False):
        self.model = DeBertaModel(model_name)
        self.extract_narrative_only = extract_narrative_only

    async def analyze_message(self, message_text: str) -> Swift799Message:
        # Prétraitement optionnel pour extraire la partie narrative
        text_to_analyze = self.extract_narrative_part(message_text) if self.extract_narrative_only else message_text
        
        # Analyser le message avec le modèle
        return await self.model.analyze_swift_message(text_to_analyze)

    def extract_narrative_part(self, message_text: str) -> str:
        """Extrait la partie narrative d'un message SWIFT."""
        
        # Recherche de la partie narrative après les en-têtes et codes
        narrative_match = re.search(r'(?:^|\n)\s*77E:?\s*(.+?)(?:\n|$)', message_text, re.DOTALL)
        if narrative_match:
            return narrative_match.group(1).strip()

        # Si pas de champ 77E, chercher après le dernier champ numéroté
        last_field_match = re.search(r'(?:^|\n)\s*\d{2}[A-Z]:?\s*(.+?)(?:\n|$)', message_text, re.DOTALL | re.RIGHTMOST)
        if last_field_match:
            return last_field_match.group(1).strip()

        # Si aucun format reconnu, retourner le texte complet
        return message_text.strip()

File: SwiftAnalyzer.Core/Swift799Analyzer.cs
using System;
using System.Collections.Generic;
using System.Text.RegularExpressions;
using System.Threading.Tasks;
using SwiftAnalyzer.Core.Models;
using SwiftAnalyzer.Core.Services.ML;
using SubTask = SwiftAnalyzer.Core.Models.SubTask;

namespace SwiftAnalyzer.Core
{
    public class Swift799Analyzer : IDisposable
    {
        private readonly DeBertaModel _model;
        private readonly bool _extractNarrativeOnly;

        public Swift799Analyzer(string modelPath, string tokenizerPath, string vocabPath, bool extractNarrativeOnly = false)
        {
            _model = new DeBertaModel(modelPath, tokenizerPath, vocabPath);
            _extractNarrativeOnly = extractNarrativeOnly;
        }

        public async Task<Swift799Message> AnalyzeMessage(string messageText)
        {
            // Prétraitement optionnel pour extraire la partie narrative
            var textToAnalyze = _extractNarrativeOnly 
                ? ExtractNarrativePart(messageText) 
                : messageText;

            var analysis = await _model.AnalyzeSwiftMessage(textToAnalyze);
            
            var message = new Swift799Message
            {
                OriginalText = messageText,
                MainTask = analysis.MainOperationType,
                HasConditionalActions = analysis.HasConditions
            };

            foreach (var subTask in analysis.SubTasks)
            {
                message.SubTasks.Add(new SubTask
                {
                    Type = subTask.Type,
                    Text = subTask.Text,
                    Confidence = subTask.Confidence,
                    RelatedTasks = subTask.RelatedTasks
                });
            }

            foreach (var timeConstraint in analysis.TimeConstraints)
            {
                if (timeConstraint.Item2 > 0.3)
                {
                    message.TimeConstraints.Add(new TimeConstraint
                    {
                        Type = timeConstraint.Item1,
                        Confidence = timeConstraint.Item2
                    });
                }
            }

            return message;
        }

        private string ExtractNarrativePart(string messageText)
        {
            // Recherche de la partie narrative après les en-têtes et codes
            var narrativeMatch = Regex.Match(messageText, @"(?:^|\n)\s*77E:?\s*(.+?)(?:\n|$)", RegexOptions.Singleline);
            if (narrativeMatch.Success)
            {
                return narrativeMatch.Groups[1].Value.Trim();
            }

            // Si pas de champ 77E, chercher après le dernier champ numéroté
            var lastFieldMatch = Regex.Match(messageText, @"(?:^|\n)\s*\d{2}[A-Z]:?\s*(.+?)(?:\n|$)", RegexOptions.Singleline | RegexOptions.RightToLeft);
            if (lastFieldMatch.Success)
            {
                return lastFieldMatch.Groups[1].Value.Trim();
            }

            // Si aucun format reconnu, retourner le texte complet
            return messageText.Trim();
        }

        public void Dispose()
        {
            _model?.Dispose();
        }
    }
}
